runtime/shared_heap.c

```diff
@@ -785,6 +811,475 @@ void caml_verify_heap(caml_domain_state *domain) {
+        /* Mark this pool as evacuating */
+        cur_pool->evacuating = 1;
+
+        if( evac_pool == NULL ) {
+          evac_pool = cur_pool;
+        }
+      }
+
+      cur_blocks += (POOL_WSIZE - POOL_HEADER_SZ) / wsize_sizeclass[sz_class];
+
+      cur_pool = cur_pool->next;
+    }
+
+    /* Reset to the first pool to allocate in to */
+    cur_pool = heap->unswept_avail_pools[sz_class];
+
+    /* The first allocating pool cannot be evacuating. Since we need enough
+    pools to satisfy the number of non-empty blocks and non-empty blocks > 0,
+    the first pool we allocate in to must not be evacuating. */
+    CAMLassert( cur_pool->evacuating == 0 );
+
+    /* There are two linked lists in the following loop. cur_pool which is
+    the head of a linked list of pools we're allocating into (the to space)
+    and evac_pool which is the head of a linked list of pools we're moving from
+    (the from space). */
+
+    /* Now we start from evac_pool and evacuate live blocks to cur_pool */
+    while( evac_pool != NULL ) {
+      value* p = (value*)((char*)evac_pool + POOL_HEADER_SZ);
+      value* end = (value*)evac_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+      value *next = NULL;
+
+      /* Every pool in the list of evacuating pools must be evacuating */
+      CAMLassert( evac_pool->evacuating == 1 );
+
+      while (p + wh <= end) {
+        header_t hd = (header_t)atomic_load_relaxed((atomic_uintnat*)p);
+
+        /* A zero header in a shared heap pool indicates an empty space */
+        if( hd != 0 ) {
+          /* Reminder: since colours have rotated, UNMARKED indicates a MARKED
+          (i.e live) block */
+          if( Has_status_hd(hd, caml_global_heap_state.UNMARKED) ) {
+            /* live block in an evacuating pool, now we allocate it in
+                to cur_pool */
+            value* new_p = cur_pool->next_obj;
+            next = (value*)new_p[1];
+            cur_pool->next_obj = next;
+
+            if( next == NULL ) {
+              /* This pool has no more free space. Move it to unswept_full_pools
+                and then advance cur_pool onwards */
+              pool* next_pool = cur_pool->next;
+
+              heap->unswept_avail_pools[sz_class] = next_pool;
+              cur_pool->next = heap->unswept_full_pools[sz_class];
+              heap->unswept_full_pools[sz_class] = cur_pool;
+
+              cur_pool = next_pool;
+
+              CAMLassert( cur_pool->evacuating == 0 );
+            }
+
+            /* Copy the block to the new location */
+            memcpy(new_p, p, Whsize_hd(hd) * sizeof(value));
+
+            /* Set first field of p to as a forwarding pointer */
+            Field(Val_hp(p), 0) = Val_hp(new_p);
+          }
+          /* We are implicitly sweeping pools in the evacuation set and thus
+             we must remember to call finalisers for Custom blocks. */
+          else if( Has_status_hd(hd, caml_global_heap_state.GARBAGE) )
```

Comment:
I assume this is unmarked blocks, as the colours have rotated? If so, worth commenting.